### DISCUSSION TITLE
Seamless ARM runners

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12512] [Ideas]
## Problem

Dagger supports multi-platform operations. The most common use of this feature is to run containers on both x86 and arm architectures. However in its standard, single-node configuration, the engine will use emulation under the hood, resulting in **poor performance**. In theory, this can be addressed by distributing execution across several machines, one for each architecture - the underlying buildkit runtime provides plumbing to do this. However in practice, configure buildkit in this way is not trivial, and integrating it into Dagger specifically may require additional, unknown work.

In a nutshell, multi-platform support in Dagger **is incomplete** until operators can seamlessly provision x86 & ARM runners.

## Solution

Provide documentation and other resources (scripts, configuration templates..) to allow operators to seamlessly run Dagger across x86 and ARM runners.

**Comment by sipsma** [comment]
Rough immediate thoughts: we could support connecting to multiple runner backends from the session and then just route the requests based on the `Platform` field in the container/directory/file ID payloads.

I think the tricky part would be that currently separate runners don't know anything about each other, so when it came time to try to push a multiplatform image you'd somehow need them to coordinate between each other to push the different platforms for the image without stomping over each other.
* I forget all the details of the image spec and registry apis, so I can't say for sure immediately how easy or hard this would be.
* depot.dev appears to support this, very curious to learn how they do it: https://depot.dev/docs/guides/arm-containers#what-about-multi-architecture-containers

There's some other tricky related problems, e.g. you might want to build some binaries for different architectures and then combine them into a single directory (with copy, mergeop, whatever). Since the runners don't know about each other, they wouldn't be able to share cache. Even remote cache doesn't currently get shared for results in the middle of a session (though that's certainly something we could investigate).

---

All of the above makes me suspect it would be worth looking into "server-side" solutions too where the backend runners handle the routing and cache sharing even though the "client-side" routing solution seems so simple on the surface. It probably fits into the larger "distributable engine" story in that respect.
